src/main/kotlin/org/rust/ide/refactoring/introduceConstant/ui.kt

```diff
@@ -0,0 +1,157 @@
+        dropHighlighter()
+    }
+
+    private fun dropHighlighter() {
+        highlighter?.dispose()
+
+//        comment?.let { c ->
+//            project.runWriteCommandAction {
+//                c.delete()
+//            }
+//        }
+    }
+}
+
+fun showInsertionChooser(
+    editor: Editor,
+    expr: RsExpr,
+    callback: (InsertionCandidate) -> Unit
+) {
+    val candidates = findInsertionCandidates(expr)
+    if (isUnitTestMode) {
+        callback(MOCK!!.chooseInsertionPoint(expr, candidates))
+    } else {
+        val highlighter = Highlighter(expr, expr.project, editor)
+        JBPopupFactory.getInstance()
+            .createPopupChooserBuilder(candidates)
+            .setRenderer(object : DefaultListCellRenderer() {
+                override fun getListCellRendererComponent(list: JList<*>?,
+                                                          value: Any,
+                                                          index: Int,
+                                                          isSelected: Boolean,
+                                                          cellHasFocus: Boolean): Component {
+                    val candidate = value as InsertionCandidate
+                    val text = candidate.description()
+                    return super.getListCellRendererComponent(list, text, index, isSelected, cellHasFocus)
+                }
+            })
+            .setItemSelectedCallback { value: InsertionCandidate? ->
+                if (value == null) return@setItemSelectedCallback
+                highlighter.onSelect(value)
+            }
+            .setTitle("Insert ${expr.text} into:")
+            .setMovable(true)
+            .setResizable(false)
+            .setRequestFocus(true)
+            .setItemChosenCallback { it?.let { callback(it) } }
+            .addListener(highlighter)
+            .createPopup()
+            .showInBestPositionFor(editor)
+    }
+}
+
+interface ExtractConstantUi {
+    fun chooseInsertionPoint(expr: RsExpr, candidates: List<InsertionCandidate>): InsertionCandidate
+}
+
+data class InsertionCandidate(val context: PsiElement, val parent: PsiElement, val anchor: PsiElement) {
+    fun description(): String = when (val element = this.context) {
+        is RsFunction -> "fn ${element.name}"
+        is RsModItem -> "mod ${element.name}"
+        is RsFile -> "file"
+        else -> error("unreachable")
+    }
+}
+
+private fun findInsertionCandidates(expr: RsExpr): List<InsertionCandidate> {
+    var parent: PsiElement = expr
+    var anchor: PsiElement = expr
+    val points = mutableListOf<InsertionCandidate>()
+
+    fun getAnchor(parent: PsiElement, anchor: PsiElement): PsiElement {
+        var found = anchor
+        while (found.parent != parent) {
+            found = found.parent
+        }
+        return found
+    }
+
+    while (parent !is RsFile) {
+        parent = parent.parent
+        when (parent) {
+            is RsFunction -> {
```

Comment:
There is a one corner case where this doesn't work: a module can be declared inside a function:
```rust
fn bar() {
    // <- constant should not be suggested here
    mod foo {
        fn foo() {
            let a = 2+2;
        }
    }
}
```
but if I declare a constant inside a function `bar`, I can't use it inside module `foo` :shrug: